File: C/Medium/3296 Minimum Number of Seconds to Make Mountain Height Zero.c
#include <math.h>

long long minNumberOfSeconds(int mountainHeight, int* workerTimes, int workerTimesSize) {
    long long l = 1;
    long long minWorkerTime = workerTimes[0];
    for (int i = 1; i < workerTimesSize; i++) {
        if (workerTimes[i] < minWorkerTime) {
            minWorkerTime = workerTimes[i];
        }
    }
    
    long long r = minWorkerTime * (long long)mountainHeight * (mountainHeight + 1) / 2;
    
    while (l < r) {
        long long mid = l + (r - l) / 2;
        long long reducedHeight = 0;
        
        for (int i = 0; i < workerTimesSize; i++) {
            reducedHeight += (long long)((-1.0 + sqrt(1.0 + 8.0 * (double)mid / workerTimes[i])) / 2.0);
        }
        
        if (reducedHeight < mountainHeight) {
            l = mid + 1;
        } else {
            r = mid;
        }
    }
    
    return l;
}

File: Java/Medium/3296 Minimum Number of Seconds to Make Mountain Height Zero.java
class Solution {
    public long minNumberOfSeconds(int mountainHeight, int[] workerTimes) {
        long l = 1;
        long minWorkerTime = workerTimes[0];
        for (int t : workerTimes) {
            minWorkerTime = Math.min(minWorkerTime, t);
        }
        
        // Worst-case upper bound assuming the fastest worker does all the work
        long r = minWorkerTime * (long) mountainHeight * (mountainHeight + 1) / 2;
        
        while (l < r) {
            long mid = l + (r - l) / 2;
            if (getReducedHeight(workerTimes, mid) < mountainHeight) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return l;
    }
    
    private long getReducedHeight(int[] workerTimes, long m) {
        long reducedHeight = 0;
        for (int w : workerTimes) {
            // Using the quadratic formula to find the max height reduced in 'm' seconds
            reducedHeight += (long) ((-1 + Math.sqrt(1 + 8.0 * (double) m / w)) / 2);
        }
        return reducedHeight;
    }
}
